lib/fs.js

```diff
@@ -453,40 +454,86 @@ function closeSync(fd) {
   handleErrorFromBinding(ctx);
 }
 
-function open(path, flags, mode, callback) {
+// This function has two possible signatures:
+// fs.open(path[, flag[, mode]], callback)
+// fs.open(path[, options], callback)
+function open(path, flag, mode, callback) {
+  let options;
+  let parents;
   path = getValidatedPath(path);
   if (arguments.length < 3) {
-    callback = flags;
-    flags = 'r';
+    callback = flag;
+    flag = 'r';
     mode = 0o666;
   } else if (typeof mode === 'function') {
     callback = mode;
     mode = 0o666;
-  } else {
-    mode = parseFileMode(mode, 'mode', 0o666);
   }
-  const flagsNumber = stringToFlags(flags);
+  // Handle case where 2nd argument is options-bag
+  if (typeof flag === 'object' && flag) {
+    options = flag;
+    flag = options.flag ?? 'r';
+    mode = options.mode ?? 0o666;
+    parents = options.parents;
+  }
+  flag = stringToFlags(flag);
+  mode = parseFileMode(mode, 'mode', 0o666);
   callback = makeCallback(callback);
 
   const req = new FSReqCallback();
-  req.oncomplete = callback;
+  req.oncomplete = (err, fd) => {
```

Comment:
If `parents` is not set, please avoid creating this closure in the first place.

Reply:
k, added conditional-branch to avoid closure if possible in commit 480eaf7d:
is below what was wanted?

```diff
diff --git a/lib/fs.js b/lib/fs.js
index e5ac3f6e01..28d3f81474 100644
--- a/lib/fs.js
+++ b/lib/fs.js
@@ -481,21 +481,25 @@ function open(path, flag, mode, callback) {
   callback = makeCallback(callback);

   const req = new FSReqCallback();
-  req.oncomplete = (err, fd) => {
-    // Lazily create parent-subdirectories if they do not exist
-    if (err?.code === 'ENOENT' && parents && (flag & O_CREAT)) {
-      mkdir(pathModule.dirname(path), { recursive: true }, (err2) => {
-        if (err2) {
-          callback(err2);
-          return;
-        }
-        // Retry open() after lazily creating parent-subdirectories
-        open(path, { flag, mode }, callback);
-      });
-      return;
-    }
-    callback(err, fd);
-  };
+  if (parents) {
+    req.oncomplete = (err, fd) => {
+      // Lazily create parent-subdirectories if they do not exist
+      if (err?.code === 'ENOENT' && parents && (flag & O_CREAT)) {
+        mkdir(pathModule.dirname(path), { recursive: true }, (err2) => {
+          if (err2) {
+            callback(err2);
+            return;
+          }
+          // Retry open() after lazily creating parent-subdirectories
+          open(path, { flag, mode }, callback);
+        });
+        return;
+      }
+      callback(err, fd);
+    };
+  } else {
+    req.oncomplete = callback;
+  }

   binding.open(pathModule.toNamespacedPath(path),
                flag,
```